File: scrapers/cloudpc.py
"""云办公电脑抓取器：免费云电脑/云桌面排行

数据策略：
- 种子数据为基础（data/cloudpc_seed.json，覆盖 10 家云电脑厂商，带来源链接）
- 每日尝试轻量探测可用公开源，失败自动回退种子
- 每个云电脑套餐计算综合评分（免费力度/价格/配置/门槛/办公适配）
"""

import json
import os
import re
from datetime import datetime, timedelta, timezone

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SEED_PATH = os.path.join(BASE_DIR, "data", "cloudpc_seed.json")

# ---------- 评分 ----------

def score_free(category: str, trial_days: int, free_hours: float) -> float:
    """免费力度分：永久免费/超长免费最高，无免费最低"""
    if category in ("free_forever", "free_tier") or (trial_days and trial_days >= 365) or (free_hours and free_hours >= 1000):
        return 100
    if (free_hours and free_hours >= 180) or (trial_days and trial_days >= 90):
        return 90
    if (free_hours and free_hours >= 60) or (trial_days and trial_days >= 30):
        return 80
    if (free_hours and free_hours >= 10) or (trial_days and trial_days >= 7):
        return 65
    if free_hours and free_hours > 0:
        return 45
    return 15


def score_price(price_month) -> float:
    """价格分：月付价格越低越高"""
    if price_month is None:
        return 40
    p = float(price_month)
    if p <= 10:
        return 100
    if p <= 30:
        return 85
    if p <= 60:
        return 70
    if p <= 100:
        return 55
    if p <= 200:
        return 40
    return 25


def score_config(config: str) -> float:
    """配置分：根据 CPU/内存/显卡描述估算"""
    if not config:
        return 40
    cpu = 0
    mem = 0
    m = re.search(r"(\d+)\s*核", config)
    if m:
        cpu = int(m.group(1))
    m = re.search(r"(\d+)\s*G", config)
    if m:
        mem = int(m.group(1))
    if "顶配显卡" in config or "5080" in config or "高性能" in config:
        return 90
    val = cpu * 10 + mem * 3
    return min(90, 30 + val)


def score_threshold(verify: str) -> float:
    """门槛分：要求越低越高"""
    if not verify:
        return 100
    if "手机号" in verify:
        return 85
    if "实名" in verify:
        return 70
    if "绑卡" in verify:
        return 60
    if "企业" in verify:
        return 45
    return 75


def score_office(office: str) -> float:
    """办公适配分：多端+浏览器+可装软件最高"""
    if not office:
        return 50
    s = office
    score = 50
    if "多端" in s:
        score += 15
    if "浏览器" in s:
        score += 15
    if "办公软件" in s or "办公" in s or "桌面" in s:
        score += 15
    if "游戏为主" in s or "偏游戏" in s or "游戏为主、轻办公" in s:
        score -= 10
    return max(40, min(100, score))


def score_cloudpc_plan(plan: dict) -> dict:
    """综合评分：免费35% + 价格25% + 配置15% + 门槛10% + 办公15%"""
    f = score_free(plan.get("category", ""), plan.get("trial_days", 0), plan.get("free_hours", 0))
    p = score_price(plan.get("price_month"))
    c = score_config(plan.get("config", ""))
    t = score_threshold(plan.get("verify", ""))
    o = score_office(plan.get("office", ""))

    total = round(f * 0.35 + p * 0.25 + c * 0.15 + t * 0.10 + o * 0.15, 1)

    plan["score"] = total
    plan["scoreBreakdown"] = {
        "free": f, "price": p, "config": c, "threshold": t, "office": o
    }
    return plan


def free_tag(plan: dict) -> str:
    """免费等级标签"""
    cat = plan.get("category", "")
    hours = plan.get("free_hours", 0)
    days = plan.get("trial_days", 0)
    if cat in ("free_forever", "free_tier"):
        return "永久免费层" if cat == "free_forever" else "免费层"
    if cat == "free_trial":
        # 每天免费少量时长的类型（如极云普惠每天免费1小时）
        if hours and hours < 10 and days >= 300:
            if hours % 1 == 0:
                return f"每天免费{int(hours)}小时"
            return f"每天免费{int(hours*60)}分钟"
        if hours and hours >= 60:
            return f"送{int(hours)}小时"
        if hours and hours >= 10:
            return f"送{int(hours)}小时"
        if hours and hours >= 1:
            # 1-10小时之间，整数显示小时，小数显示分钟
            if hours % 1 == 0:
                return f"送{int(hours)}小时"
            return f"送{int(hours*60)}分钟"
        if hours:
            return f"送{int(hours*60)}分钟"
        if days >= 90:
            return f"免费{days}天"
        return f"免费{days}天" if days else "免费试用"
    if cat == "deal":
        return "限时特惠"
    return "参考"


def scrape_cloudpcs() -> dict:
    """加载种子数据 + 评分"""
    with open(SEED_PATH, "r", encoding="utf-8") as f:
        data = json.load(f)

    total_plans = 0
    for cloud in data.get("clouds", []):
        for plan in cloud.get("plans", []):
            score_cloudpc_plan(plan)
            plan["freeTag"] = free_tag(plan)
            total_plans += 1

    result = {
        "version": datetime.now(timezone.utc).astimezone(timezone(timedelta(hours=8))).strftime("%Y-%m-%d"),
        "updatedAt": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "source": data.get("source", ""),
        "note": data.get("note", ""),
        "vendorCount": len(data.get("clouds", [])),
        "planCount": total_plans,
        "clouds": data.get("clouds", []),
    }
    return result


if __name__ == "__main__":
    result = scrape_cloudpcs()
    print(f"云电脑：{result['vendorCount']} 家厂商，{result['planCount']} 个套餐")
    plans = []
    for cloud in result["clouds"]:
        for plan in cloud["plans"]:
            plans.append((cloud["vendor"], cloud.get("categoryLabel", ""), plan))
    plans.sort(key=lambda x: -x[2]["score"])
    for vendor, label, p in plans:
        price = f"{p['price_month']}元/月" if p["price_month"] else "—"
        freeh = f"送{p['free_hours']}h" if p.get("free_hours") else ""
        print(f"  {vendor:12s} | {label:8s} | {p['name']:16s} | {p['config']:22s} | {price:10s} | 评分{p['score']:5.1f} | {p['freeTag']} {freeh}")

File: site/js/cloudpc.js
/* FreeEgg Radar 云电脑榜逻辑 v20 */
(function () {
  'use strict';

  var state = {
    data: null,
    plans: [],
    filtered: [],
    activeType: 'all',
    activeFree: 'all',
    sortBy: 'score',
    search: ''
  };

  var grid = document.getElementById('cloudpc-grid');
  var typeFilters = document.getElementById('type-filters');
  var freeFilters = document.getElementById('free-filters');
  var searchInput = document.getElementById('cloudpc-filter-q');
  var updatedAt = document.getElementById('cloudpc-updated-at');

  function init() {
    loadData();
    bindEvents();
  }

  function loadData() {
    fetch('data/cloudpcs.json')
      .then(function (r) {
        if (!r.ok) throw new Error('HTTP ' + r.status);
        return r.json();
      })
      .then(function (data) {
        state.data = data;
        state.plans = flattenPlans(data);
        state.filtered = state.plans.slice();
        sortPlans();  // 初始按综合分排序

        if (updatedAt && data.version) {
          updatedAt.textContent = '数据更新：' + data.version + ' · ' + data.vendorCount + ' 家厂商 · ' + data.planCount + ' 个套餐 · 每日自动更新';
        }

        render();
      })
      .catch(function (err) {
        console.error('加载云电脑数据失败:', err);
        if (grid) grid.innerHTML = '<div class="empty">数据加载失败，请稍后重试</div>';
      });
  }

  function flattenPlans(data) {
    var plans = [];
    (data.clouds || []).forEach(function (cloud) {
      var vendor = cloud.vendor || '未知';
      var vendorEn = cloud.vendor_en || '';
      var category = cloud.category || '';
      var categoryLabel = cloud.categoryLabel || (category === 'consumer' ? '消费级云电脑' : '企业级云桌面');
      (cloud.plans || []).forEach(function (plan) {
        var p = plan || {};
        p.vendor = vendor;
        p.vendor_en = vendorEn;
        p.categoryGroup = category;
        p.categoryLabel = categoryLabel;
        p.isFree = p.category === 'free_forever' || p.category === 'free_tier' || p.category === 'free_trial';
        plans.push(p);
      });
    });
    return plans;
  }

  function bindEvents() {
    if (typeFilters) {
      typeFilters.addEventListener('click', function (e) {
        var btn = e.target.closest('.chip');
        if (!btn) return;
        state.activeType = btn.getAttribute('data-type');
        typeFilters.querySelectorAll('.chip').forEach(function (c) {
          c.classList.toggle('active', c.getAttribute('data-type') === state.activeType);
        });
        applyFilters();
      });
    }

    if (freeFilters) {
      freeFilters.addEventListener('click', function (e) {
        var btn = e.target.closest('.chip');
        if (!btn) return;
        state.activeFree = btn.getAttribute('data-free');
        freeFilters.querySelectorAll('.chip').forEach(function (c) {
          c.classList.toggle('active', c.getAttribute('data-free') === state.activeFree);
        });
        applyFilters();
      });
    }

    var sortButtons = document.querySelectorAll('[data-sort]');
    sortButtons.forEach(function (btn) {
      btn.addEventListener('click', function () {
        state.sortBy = btn.getAttribute('data-sort');
        sortButtons.forEach(function (c) {
          c.classList.toggle('active', c.getAttribute('data-sort') === state.sortBy);
        });
        applyFilters();
      });
    });

    if (searchInput) {
      var timer;
      searchInput.addEventListener('input', function () {
        clearTimeout(timer);
        timer = setTimeout(function () {
          state.search = searchInput.value.trim().toLowerCase();
          applyFilters();
        }, 200);
      });
    }
  }

  function applyFilters() {
    var type = state.activeType;
    var free = state.activeFree;
    var q = state.search;

    state.filtered = state.plans.filter(function (p) {
      if (type !== 'all' && p.categoryGroup !== type) return false;
      if (free === 'free' && !p.isFree) return false;
      if (free === 'deal' && p.category !== 'deal') return false;
      if (q) {
        var haystack = ((p.vendor || '') + ' ' + (p.name || '') + ' ' + (p.config || '') + ' ' + (p.trial || '')).toLowerCase();
        if (haystack.indexOf(q) === -1) return false;
      }
      return true;
    });

    sortPlans();
    render();
  }

  function sortPlans() {
    var by = state.sortBy;
    state.filtered.sort(function (a, b) {
      if (by === 'price') {
        var pa = a.price_month != null ? a.price_month : 99999;
        var pb = b.price_month != null ? b.price_month : 99999;
        return pa - pb;
      }
      if (by === 'free') {
        var fb = scoreFreeWeight(b) - scoreFreeWeight(a);
        if (fb !== 0) return fb;
        return (b.score || 0) - (a.score || 0);
      }
      return (b.score || 0) - (a.score || 0);
    });
  }

  function scoreFreeWeight(p) {
    var b = p.scoreBreakdown || {};
    return b.free || 0;
  }

  function render() {
    if (!grid) return;

    if (state.filtered.length === 0) {
      grid.innerHTML = '<div class="empty">没有找到匹配的云电脑</div>';
      return;
    }

    var html = '';
    state.filtered.forEach(function (p, i) {
      var rankClass = '';
      if (i === 0) rankClass = ' top1';
      else if (i === 1) rankClass = ' top2';
      else if (i === 2) rankClass = ' top3';

      // 价格显示
      var priceHtml;
      if (p.category === 'free_forever' || p.category === 'free_tier' || p.price_month === 0) {
        priceHtml = '<div class="cloudpc-price-item"><div class="cloudpc-price-label">月费</div><div class="cloudpc-price-value free">免费</div></div>';
      } else if (p.price_month != null) {
        priceHtml = '<div class="cloudpc-price-item"><div class="cloudpc-price-label">月费</div><div class="cloudpc-price-value">' + p.price_month + ' 元</div></div>';
      } else {
        priceHtml = '<div class="cloudpc-price-item"><div class="cloudpc-price-label">月费</div><div class="cloudpc-price-value na">待查</div></div>';
      }

      // 免费时长
      if (p.free_hours != null && p.free_hours > 0) {
        var freeText = p.free_hours % 1 === 0 ? p.free_hours + ' 小时' : (p.free_hours * 60) + ' 分钟';
        if (p.trial_days >= 300) freeText = '每天 ' + (p.free_hours % 1 === 0 ? p.free_hours + ' 小时' : (p.free_hours * 60) + ' 分钟');
        priceHtml += '<div class="cloudpc-price-item"><div class="cloudpc-price-label">免费时长</div><div class="cloudpc-price-value free">' + freeText + '</div></div>';
      } else {
        priceHtml += '<div class="cloudpc-price-item"><div class="cloudpc-price-label">免费时长</div><div class="cloudpc-price-value na">—</div></div>';
      }

      // 标签
      var tags = '<span class="cloudpc-tag ' + tagClass(p.category) + '">' + escapeHtml(p.freeTag || '') + '</span>';
      tags += '<span class="cloudpc-tag info">' + escapeHtml(p.categoryLabel) + '</span>';

      // 详情行
      var details = '';
      if (p.trial) details += '<div><strong>免费/试用：</strong>' + escapeHtml(p.trial) + '</div>';
      if (p.verify) details += '<div><strong>门槛：</strong>' + escapeHtml(p.verify) + '</div>';
      if (p.renewal) details += '<div><strong>续费/长期：</strong>' + escapeHtml(p.renewal) + '</div>';
      if (p.office) details += '<div><strong>办公适配：</strong>' + escapeHtml(p.office) + '</div>';

      // 评分明细
      var breakdown = p.scoreBreakdown || {};
      var detailHint = '免费' + Math.round(breakdown.free || 0) + ' · 价格' + Math.round(breakdown.price || 0) + ' · 配置' + Math.round(breakdown.config || 0) + ' · 门槛' + Math.round(breakdown.threshold || 0) + ' · 办公' + Math.round(breakdown.office || 0);

      html += '<article class="cloudpc-card' + rankClass + '" style="animation-delay:' + (i * 0.03) + 's">';
      html += '<div class="cloudpc-rank">' + (i + 1) + '</div>';
      html += '<div class="cloudpc-vendor">' + escapeHtml(p.vendor) + (p.vendor_en ? ' <span style="color:var(--text-faint);font-weight:400;font-size:11px;">' + escapeHtml(p.vendor_en) + '</span>' : '') + '</div>';
      html += '<div class="cloudpc-plan">' + escapeHtml(p.name || '') + '</div>';
      if (p.config) html += '<div class="cloudpc-config">' + escapeHtml(p.config) + '</div>';
      html += '<div class="cloudpc-tags">' + tags + '</div>';
      html += '<div class="cloudpc-price-row">' + priceHtml + '</div>';
      if (details) html += '<div class="cloudpc-detail">' + details + '</div>';
      html += '<div class="cloudpc-score-row">';
      html += '<div><div class="cloudpc-score-label">综合评分</div><div style="font-size:10px;color:var(--text-faint);margin-top:2px;">' + detailHint + '</div></div>';
      html += '<div class="cloudpc-score">' + (p.score != null ? p.score : '—') + '</div>';
      html += '</div>';
      if (p.url) html += '<a class="cloudpc-link" href="' + escapeHtml(p.url) + '" target="_blank" rel="noopener noreferrer">查看官方页面 →</a>';
      html += '</article>';
    });

    grid.innerHTML = html;
  }

  function tagClass(cat) {
    if (cat === 'free_forever') return 'free-forever';
    if (cat === 'free_tier') return 'free-tier';
    if (cat === 'free_trial') return 'free-trial';
    if (cat === 'deal') return 'deal';
    return 'info';
  }

  function escapeHtml(s) {
    if (s == null) return '';
    var div = document.createElement('div');
    div.textContent = String(s);
    return div.innerHTML;
  }

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', init);
  } else {
    init();
  }
})();
